client/task_runner.go

```diff
@@ -356,25 +423,287 @@ func (r *TaskRunner) validateTask() error {
+// token exists
+type tokenFuture struct {
+	waiting []chan struct{}
+	token   string
+	set     bool
+	m       sync.Mutex
+}
+
+// NewTokenFuture returns a new token future without any token set
+func NewTokenFuture() *tokenFuture {
+	return &tokenFuture{}
+}
+
+// Wait returns a channel that can be waited on. When this channel unblocks, a
+// valid token will be available via the Get method
+func (f *tokenFuture) Wait() <-chan struct{} {
+	f.m.Lock()
+	defer f.m.Unlock()
+
+	c := make(chan struct{})
+	if f.set {
+		close(c)
+		return c
+	}
+
+	f.waiting = append(f.waiting, c)
+	return c
+}
+
+// Set sets the token value and unblocks any caller of Wait
+func (f *tokenFuture) Set(token string) *tokenFuture {
+	f.m.Lock()
+	defer f.m.Unlock()
+
+	f.set = true
+	f.token = token
+	for _, w := range f.waiting {
+		close(w)
+	}
+	f.waiting = nil
+	return f
+}
+
+// Clear clears the set vault token.
+func (f *tokenFuture) Clear() *tokenFuture {
+	f.m.Lock()
+	defer f.m.Unlock()
+
+	f.token = ""
+	f.set = false
+	return f
+}
+
+// Get returns the set Vault token
+func (f *tokenFuture) Get() string {
+	f.m.Lock()
+	defer f.m.Unlock()
+	return f.token
+}
+
+// vaultManager should be called in a go-routine and manages the derivation,
+// renewal and handling of errors with the Vault token. The optional parameter
+// allows setting the initial Vault token. This is useful when the Vault token
+// is recovered off disk.
+func (r *TaskRunner) vaultManager(token string) {
+	// updatedToken lets us store state between loops. If true, a new token
+	// has been retrieved and we need to apply the Vault change mode
+	var updatedToken bool
+
+OUTER:
+	for {
+		// Check if we should exit
+		select {
+		case <-r.waitCh:
+			return
+		default:
+		}
+
+		// Clear the token
+		r.vaultFuture.Clear()
+
+		// Check if there already is a token which can be the case for
+		// restoring the TaskRunner
+		if token == "" {
+			// Get a token
+			var ok bool
+			token, ok = r.deriveVaultToken()
+			if !ok {
+				// We are shutting down
+				return
+			}
+
+			// Write the token to disk
+			if err := r.writeToken(token); err != nil {
+				e := fmt.Errorf("failed to write Vault token to disk")
+				r.logger.Printf("[ERR] client: %v for task %v on alloc %q: %v", e, r.task.Name, r.alloc.ID, err)
+				r.Kill("vault", e.Error())
+				return
+			}
+		}
+
+		// Start the renewal process
+		renewCh, err := r.vaultClient.RenewToken(token, 30)
+
+		// An error returned means the token is not being renewed
+		if err != nil {
+			r.logger.Printf("[ERR] client: failed to start renewal of Vault token for task %v on alloc %q: %v", r.task.Name, r.alloc.ID, err)
+			token = ""
+			goto OUTER
+		}
+
+		// The Vault token is valid now, so set it
+		r.vaultFuture.Set(token)
+
+		if updatedToken {
+			switch r.task.Vault.ChangeMode {
+			case structs.VaultChangeModeSignal:
+				s, err := signals.Parse(r.task.Vault.ChangeSignal)
```

Comment:
Yep an upcoming PR will make drivers fingerprint the signals it can handle, and do server side constraints and then push the logic into `validateTask`. But this PR was already big